File: translator/src/rendering/codes/loop.py
from typing import Optional

from .code import Code
from src.symbols import Symbols
from src.templates import Templates


class LoopCode(Code):
    def __init__(self, symbols: Symbols, templates: Templates):
        super().__init__(
            main_template=templates.LOOP,
            secondary_template=templates.DECREASE_REF_COUNT,
            main_data={
                "type": symbols.OBJECT_TYPE,
            },
            secondary_data={
                "func": symbols.DECREASE_REF_COUNT,
            },
        )

    def update_data(
        self,
        var: Optional[str] = None,
        pre_body: Optional[str] = None,
        post_body: Optional[str] = None,
        test_body: Optional[str] = None,
        test_value: Optional[str] = None,
        true_test_body: Optional[str] = None,
        true_test_var: Optional[str] = None,
        false_test_body: Optional[str] = None,
    ) -> None:
        self._update_main_data(
            var=var,
            pre_body=pre_body,
            post_body=post_body,
            test_body=test_body,
            test_value=test_value,
            true_test_body=true_test_body,
            true_test_var=true_test_var,
            false_test_body=false_test_body,
        )
        self._update_secondary_data(var=var)

File: translator/code_templates/loop.c
{{ type }} {{ var }};
{{ pre_body }}
while (1) {
{% set tab = "\t" %}
{% set test_body_lines = test_body.split("\n") if test_body != "" else [] %}{% for l in test_body_lines %}{{ tab }}{{ l }}
{% endfor %}
{{ tab }}if ({{ test_value }}) {
{% set true_test_body_lines = true_test_body.split("\n") if true_test_body != "" else [] %}{% for l in true_test_body_lines %}{{ tab }}{{ tab }}{{ l }}
{% endfor %}
{{ tab }}{{ tab }}{{ var }} = {{ true_test_var }};
{{ tab }}{{ tab }}break;
{{ tab }}}
{% set false_test_body_lines = false_test_body.split("\n") if false_test_body != "" else [] %}{% for l in false_test_body_lines %}{{ tab }}{{ l }}
{% endfor %}
}
{{ post_body }}
